File: LeetCode/Insert Interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval> &intervals, Interval newInterval) {
        vector<Interval> ans;
        int n = intervals.size(), k = 0;
        while (k < n){
            if (intervals[k].end < newInterval.start){
                ans.push_back(intervals[k]);
                k++;
            }else{
                break;
            }
        }
        while (k < n && !(intervals[k].end < newInterval.start || intervals[k].start > newInterval.end)){
            newInterval.start = min(intervals[k].start , newInterval.start);
            newInterval.end = max(intervals[k].end , newInterval.end);
            k++;
        }
        ans.push_back(newInterval);
        while (k < n){
            ans.push_back(intervals[k]);
            k++;
        }
        return ans;
    }
};

File: LeetCode/Insert Interval.py
# Definition for an interval.
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

class Solution:
    # @param intervals, a list of Intervals
    # @param newInterval, a Interval
    # @return a list of Interval
    def insert(self, intervals, newInterval):
        ans = []
        # print newInterval.start,newInterval.end
        flag = 0
        for item in intervals:
            # print item.start, item.end
            if item.start < newInterval.start and item.end >= newInterval.start:
                newInterval.start = item.start
            if item.end > newInterval.end and item.start <= newInterval.end:
                newInterval.end = item.end
            if newInterval.start > item.end or newInterval.end < item.start:
                if flag = 0:
                    ans.append(newInterval)
                    flag = 1
                ans.append(item)
        return ans

if __name__ == '__main__':
    s = Solution()
    i1 = Interval(1,2)
    i2 = Interval(3,5)
    i3 = Interval(6,7)
    i4 = Interval(8,10)
    i5 = Interval(12,16)
    intervals = [i1, i2, i3, i4, i5]
    newInterval = Interval(4,9)
    # ans = s.insert(intervals, newInterval)
    ans = s.insert([Interval(1,4)], newInterval)
    for item in ans:
        print (item.start,item.end)